Now `TypeInfo.get_method` also returns `Decorator` nodes

sobolevn opened the pull request:
Closes #10409

JukkaL commented on test-data/unit/check-classes.test:
```diff
@@ -2617,6 +2617,27 @@ b = a.bar
+b = a.bar
+[builtins fixtures/tuple.pyi]
+[out]
+main:16: error: Incompatible types in assignment (expression has type "A", variable has type "B")
```
Style nit: test cases will be easier to read and maintain if messages are introduced using `# E: <message:` (or `# N: ...`).

JukkaL commented on test-data/unit/check-classes.test:
```diff
@@ -2617,6 +2617,27 @@ b = a.bar
+T = TypeVar('T', bound=Callable)
+
+def decorator(f: T) -> T:
+    return f
```
Test a decorator that changes the signature of the method. I this might not work correctly right now (for example, return plain `Callable` -> it should turn the attribute type to `Any`. If this is an issue, you'll probably need to get the method type from the decorator node.

sobolevn replied:
It is working correctly, no need to change that. Thanks for the idea 👍

sobolevn replied:
Nope, I was wrong. I've misinterpreted your words in an opposite way.

Example of test that passes, but should not: https://github.com/python/mypy/pull/11150/commits/28d950d10b39e9c09865e7d4f62ab437152ba238#diff-1c26dc582ed5f36723414080dfb646bc71cf20736b73ae1ccff9b90a5d3befafR2635-R2638

Right now this happens:

```python
from typing import Callable, TypeVar

T = TypeVar('T', bound=Callable)

def decorator(f: Callable[['A', str], 'A']) -> Callable[['A', str], int]:
    def inner(self: 'A', x: str) -> int:
        return f(self, x).num
    return inner

class A:
    num = 1

    @decorator
    def __getattr__(self, x: str) -> 'A':
        return A()

a = A()

reveal_type(a.a)  # Revealed type is "ex.A"
# In runtime: 1
```

So, yeah. I need to get type from `Decorator`, not just its node.

JukkaL commented on test-data/unit/check-classes.test:
```diff
@@ -2644,6 +2665,27 @@ b = a.bar
+b = a.bar
+[builtins fixtures/tuple.pyi]
+[out]
+main:16: error: Incompatible types in assignment (expression has type "A", variable has type "B")
```
Similar to above.

JukkaL commented on mypy/checkmember.py:
```diff
@@ -383,7 +383,18 @@ def analyze_member_var_access(name: str,
-                method = info.get_method(method_name)
+                method_node = info.get(method_name)
+
+                # We cannot use `.get_method()` here, because it only returns
```
Would it be better to fix `get_method` to also work for decorated methods? The same issue could come up elsewhere.

sobolevn replied:
It is used in quite a lot of places right now:

```
» grep -o '.get_method(' -r mypy        
mypy/checkmember.py:.get_method(
mypy/checkmember.py:.get_method(
mypy/checkmember.py:.get_method(
mypy/checkmember.py:.get_method(
mypy/typeshed/stdlib/@python2/urllib2.pyi: get_method(
mypy/typeshed/stdlib/urllib/request.pyi: get_method(
mypy/nodes.py: get_method(
mypy/checker.py:.get_method(
mypy/checker.py:.get_method(
mypy/checker.py:.get_method(
mypy/subtypes.py:.get_method(
mypy/subtypes.py:.get_method(
mypy/subtypes.py:.get_method(
mypy/subtypes.py:.get_method(
```

We can probably ignore 2 `mypy/typeshed` usages, which gives us 12 total.
I will try to push an experimental change with re-worked `get_method()` to see what happens.

sobolevn replied:
A lot of tests fail locally for this change: https://github.com/python/mypy/pull/11150/commits/c9674863a6255d62ada03b0b87f945c5b52f4d72 Let's wait for the CI to finish.

JukkaL reviewed (commented):
Thanks for the PR! Left a few comments and a question.

sobolevn commented:
This commit just addresses the code review: https://github.com/python/mypy/pull/11150/commits/28d950d10b39e9c09865e7d4f62ab437152ba238

Now, I am going to change how `.get_method()` is defined. Let's see what will happen!

sobolevn commented:
I am running self-check with the only change: `def get_method(self) -> Union[FuncBase, Decorator, None]`

```
» python runtests.py self
run self: ['/Users/sobolev/Desktop/mypy/.venv/bin/python', '-m', 'mypy', '--config-file', 'mypy_self_check.ini', '-p', 'mypy']
mypy/subtypes.py:618: error: Item "Decorator" of "Union[FuncBase, Decorator]" has no
attribute "is_property"  [union-attr]
            if method.is_property:
               ^
mypy/subtypes.py:623: error: Argument 1 to "find_node_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "Union[Var, FuncBase]"  [arg-type]
            return find_node_type(method, itype, subtype)
                                  ^
mypy/subtypes.py:644: error: Argument 1 to "find_node_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "Union[Var, FuncBase]"  [arg-type]
    ...         getattr_type = get_proper_type(find_node_type(method, itype, subtyp...
                                                              ^
mypy/subtypes.py:667: error: Item "Decorator" of "Union[FuncBase, Decorator]" has no
attribute "is_property"  [union-attr]
            if method.is_property:
               ^
mypy/checkmember.py:199: error: Item "Decorator" of "Union[FuncBase, Decorator]" has
no attribute "is_property"  [union-attr]
            if method.is_property:
               ^
mypy/checkmember.py:205: error: Argument 1 to "function_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "FuncBase"  [arg-type]
            signature = function_type(method, mx.builtin_type('builtins.function...
                                      ^
mypy/checkmember.py:216: error: Item "Decorator" of "Union[FuncBase, Decorator]" has
no attribute "is_class"  [union-attr]
    ...     signature = check_self_arg(signature, dispatched_type, method.is_class,
                                                                   ^
mypy/checkmember.py:218: error: Item "Decorator" of "Union[FuncBase, Decorator]" has
no attribute "is_class"  [union-attr]
    ...gnature = bind_self(signature, mx.self_type, is_classmethod=method.is_class)
                                                                   ^
mypy/checkmember.py:392: error: Argument 1 to "function_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "FuncBase"  [arg-type]
                        function = function_type(method, mx.builtin_type('builti...
                                                 ^
mypy/checkmember.py:409: error: Argument 1 to "function_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "FuncBase"  [arg-type]
                    setattr_func = function_type(setattr_meth, mx.builtin_type('...
                                                 ^
mypy/checkmember.py:479: error: Argument 1 to "function_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "FuncBase"  [arg-type]
        function = function_type(dunder_get, builtin_type('builtins.function'))
                                 ^
mypy/checker.py:3125: error: Argument 1 to "function_type" has incompatible type
"Union[FuncBase, Decorator]"; expected "FuncBase"  [arg-type]
            function = function_type(dunder_set, self.named_type('builtins.funct...
                                     ^
mypy/checker.py:6042: error: Never apply isinstance() to unexpanded types; use
mypy.types.get_proper_type() first  [misc]
                if isinstance(method.type, Overloaded):
                   ^
mypy/checker.py:6042: note: If you pass on the original type after the check, always use its unexpanded version
Found 13 errors in 3 files (checked 164 source files)

FAILED: self
```

Looks doable! 👍

sobolevn commented:
This is now pretty big: 
1. I've changed all parts where `.get_method()` was used to work with decorators, including
  - Protocol checks with `find_member`
  - `is_untyped_decorator` function in `checker`
  - `checkmember`
2. I've changed how type was inferenced from decorators, so for example decorating `x.__getattr__` actually changes what type is inferred for `x.attr`
3. I've added new tests for each individual change, improved some existing ones

sobolevn commented on mypy/checkmember.py:
```diff
@@ -384,29 +387,46 @@ def analyze_member_var_access(name: str,
+                    else:
+                        result = getattr_type
+
+                    # Call the attribute hook before returning.
```
Now plugin hooks are also triggered for non `callable` types as well.

sobolevn commented:
Ok, looks like `mypyc` also need to be fixed. It is failing right now https://github.com/python/mypy/runs/3673913236

sobolevn commented:
`mypyc` didn't like this code:

```python
if isinstance(node, (FuncBase, Decorator)):
    return node
```

But, it accepted this code:

```python
if isinstance(node, FuncBase):
    return node
elif isinstance(node, Decorator):
    return node
```

Seems like a bug in `mypyc` to me.

JukkaL commented:
I think I found a reproducer for the mypyc issue: https://github.com/mypyc/mypyc/issues/892

sobolevn commented:
@JukkaL awesome! Thanks! Is there anything else left to do here?

sobolevn commented:
Rebased on `master`. @JukkaL is there anything left to do on this one? 🤔

sobolevn commented:
@JukkaL any chance to get this merged? 🙂

sobolevn commented:
@JukkaL rebased once again 🙂 

I will also friendly ping @hauntsaninja as well.

97littleleaf11 reviewed (commented):
Using `MemberContext` in `analyze_descriptor_access` is a good refactoring!

JukkaL commented on mypy/checker.py:
```diff
@@ -3190,9 +3194,15 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
-
-        function = function_type(dunder_set, self.named_type('builtins.function'))
-        bound_method = bind_self(function, attribute_type)
+        if isinstance(dunder_set, Decorator):
```
Code quite similar to this if/else gets repeated several times. I wonder if it would be possible to abstract it into a helper function, for example? If this sounds like a feasible idea, it could be a nice follow-up PR.

JukkaL reviewed (approved):
Thanks for the PR, this looks solid! I like how this fixes many related issues.

sobolevn commented:
@JukkaL thank you! I will think about a helper function 🙂
